src/rust/engine/fs/store/src/local.rs

```diff
@@ -453,6 +516,90 @@ impl ByteStore {
     }
     Ok(digests)
   }
+
+  async fn rename_temp_large_file(
+    &self,
+    src: NamedTempFile,
+    digest: Digest,
+  ) -> Result<Digest, String> {
+    let dest = self.large_file_path(digest);
+    tokio::fs::rename(&src, dest.clone())
+      .await
+      .map_err(|e| format!("Error while renaming: {e}."))?;
+    tokio::fs::set_permissions(&dest, std::fs::Permissions::from_mode(0o555))
+      .await
+      .map_err(|e| format!("Failed to set permissions for {dest:?}: {e}"))?;
+    Ok(digest)
+  }
+
+  async fn make_temp_dest(&self, digest: Digest) -> Result<NamedTempFile, String> {
+    let dest = self.large_file_path(digest);
+    if !dest.parent().unwrap().exists() {
+      // Throway the result as a way of not worrying about race conditions.
+      // If there was an error we'll fail later.
+      let _ = tokio::fs::create_dir_all(dest.parent().unwrap()).await;
+    }
+    // Make the tempdir in the same dir as the final file so that materializing the final file doesn't
+    // have to worry about parent dirs.
+    NamedTempFile::new_in(dest.parent().unwrap())
+      .map_err(|e| format!("Failed to create temporary file for large file: {e}"))
+  }
+
+  async fn store_large_bytes(
+    &self,
+    src: Bytes,
+    mut digest: Option<Digest>,
+  ) -> Result<Digest, String> {
+    if digest.is_none() {
+      digest =
+        Some(hash(&mut src.clone().reader()).map_err(|e| format!("Failed to hash bytes: {e}"))?);
+    }
+    let temp_dest = self.make_temp_dest(digest.unwrap()).await?;
+    tokio::fs::write(temp_dest.path(), src.clone())
+      .await
+      .map_err(|e| format!("Failed to copy from {src:?} or store in {temp_dest:?}: {e:?}"))?;
+
+    self
+      .rename_temp_large_file(temp_dest, digest.unwrap())
+      .await
+  }
+
+  async fn store_large_file(
+    &self,
+    src: PathBuf,
+    digest: Digest,
+    data_is_immutable: bool,
+  ) -> Result<(), String> {
+    let temp_dest = self.make_temp_dest(digest).await?;
+
+    let mut attempts = 0;
+    loop {
+      let should_retry = !verified_copy(
```

Comment:
Thinking to https://github.com/pantsbuild/pants/pull/18054 (when I get back to it...), it'll be possible for the remote store to write directly to a sensible temp file, and thus that file can be renamed (or hard-linked) without a separate copy pass. This can presumably be refactored as required for that, but thinking along those lines makes me wonder if there's also other places that generate known-'temporary' files that can be renamed/hard-linked instead of doing a full copy?

Reply:
> Thinking to #18054 (when I get back to it...), it'll be possible for the remote store to write directly to a sensible temp file, and thus that file can be renamed (or hard-linked) without a separate copy pass. This codepath can presumably be refactored as required for that, but thinking along those lines makes me wonder if there's also other places that generate known-'temporary' files that can be renamed/hard-linked instead of doing a full copy?

Yea, definitely.

There _should_ be more codepaths using temporary files, but file downloading (as one example) slurps stuff into memory instead.

Reply:
done :smile: